pkg/domain/infra/abi/quadlet.go

```diff
@@ -0,0 +1,708 @@
+	}
+
+	quadletDirs := systemdquadlet.GetUnitDirs(rootless.IsRootless())
+	for _, dir := range quadletDirs {
+		testPath := filepath.Join(dir, name)
+		if _, err := os.Stat(testPath); err != nil {
+			if !errors.Is(err, fs.ErrNotExist) {
+				return "", fmt.Errorf("cannot stat quadlet at path %q: %w", testPath, err)
+			}
+			continue
+		}
+		return testPath, nil
+	}
+	return "", fmt.Errorf("could not locate quadlet %q in any supported quadlet directory", name)
+}
+
+func (ic *ContainerEngine) QuadletPrint(ctx context.Context, quadlet string) (string, error) {
+	quadletPath, err := getQuadletByName(quadlet)
+	if err != nil {
+		return "", err
+	}
+
+	contents, err := os.ReadFile(quadletPath)
+	if err != nil {
+		return "", fmt.Errorf("reading quadlet %q contents: %w", quadletPath, err)
+	}
+
+	return string(contents), nil
+}
+
+func (ic *ContainerEngine) QuadletRemove(ctx context.Context, quadlets []string, options entities.QuadletRemoveOptions) (*entities.QuadletRemoveReport, error) {
+	report := entities.QuadletRemoveReport{
+		Errors:  make(map[string]error),
+		Removed: []string{},
+	}
+	removeList := []string{}
+	reverseMap, appMap, err := buildAppMap(systemdquadlet.GetInstallUnitDirPath(rootless.IsRootless()))
+	if err != nil {
+		return nil, fmt.Errorf("unable to build app map: %w", err)
+	}
+	quadletsRe := []string{}
+	// Process all `.app` files in arguments, if `.app` file
+	// is found then expand it to its respective quadlet files
+	// and remove it from the processing list.
+	for _, quadlet := range quadlets {
+		// Most likely this is an app
+		if strings.HasPrefix(quadlet, ".") && strings.HasSuffix(quadlet, ".app") {
+			files, ok := appMap[quadlet]
+			// Add all files of this application in to-be removed list.
+			if ok {
+				for _, file := range files {
+					if !systemdquadlet.IsExtSupported(file) {
+						removeList = append(removeList, file)
+					} else {
+						quadletsRe = append(quadletsRe, file)
+					}
+				}
+			}
+		} else {
+			quadletsRe = append(quadletsRe, quadlet)
+		}
+	}
+	quadlets = quadletsRe
+	allQuadletPaths := make([]string, 0, len(quadlets))
+	allServiceNames := make([]string, 0, len(quadlets))
+	runningQuadlets := make([]string, 0, len(quadlets))
+	serviceNameToQuadletName := make(map[string]string)
+	needReload := false
+
+	// Early escape: if 0 quadlets are requested, bail immediately without error.
+	if len(quadlets) == 0 && !options.All {
+		return nil, errors.New("must provide at least 1 quadlet to remove")
+	}
+
+	// Is systemd available to the current user?
+	// We cannot proceed if not.
+	conn, err := systemd.ConnectToDBUS()
+	if err != nil {
+		return nil, fmt.Errorf("connecting to systemd dbus: %w", err)
+	}
+
+	if options.All {
+		allQuadlets := getAllQuadletPaths()
+		quadlets = allQuadlets
+	}
+
+	for i := 0; i < len(quadlets); i++ {
+		quadlet := quadlets[i]
+		quadletPath, err := getQuadletByName(quadlet)
+		if options.All {
+			quadletPath = quadlet
```

Comment:
This doesn't seem right. Why are we even doing `getQuadletByName` when `options.All` is set?